src/coreclr/jit/inductionvariableopts.cpp

```diff
@@ -1885,10 +2001,15 @@ bool StrengthReductionContext::StaysWithinManagedObject(ArrayStack<CursorInfo>*
     for (int i = 0; i < cursors->Height(); i++)
     {
         CursorInfo& cursor = cursors->BottomRef(i);
-        GenTree*    parent = cursor.Tree->gtGetParent(nullptr);
-        if ((parent != nullptr) && parent->OperIs(GT_ARR_ADDR))
+        GenTree*    cur    = cursor.Tree;
+        while ((cur != nullptr) && !cur->OperIs(GT_ARR_ADDR))
+        {
+            cur = cur->gtGetParent(nullptr);
+        }
```

Comment:
This whole syntactic treatment of `GT_ARR_ADDR` to figure out what address ranges are inside arrays is unfortunate. It's pretty fragile and leads e.g. to #108706 not just working out properly. Not sure how to improve this much however -- maybe we can keep some breadcrumbs about known managed ranges during the cursor advancement and query it any time we need to prove something here.

Reply:
For advanced loop opts (parallelization, vectorization) it's not uncommon to create a "raised up" representation that describes the array element as an affine function of the IVs, so maybe we can look into something like this?

Reply:
Sounds like that could be a direction to go, although I guess it would mean that we go from `INDEX_ADDR` to expanded nodes in morph and then back to some equivalent of `INDEX_ADDR` as part of these opts. But maybe that's fine (and perhaps the structure could be recovered on VNs, similar to the comment on `ParseArrayAddress` suggests).